api/graphite.go

```diff
@@ -0,0 +1,350 @@
+	}
+
+	reqs, err = alignRequests(reqs, s.MemoryStore.AggSettings())
+	if err != nil {
+		log.Error(3, "HTTP Render alignReq error: %s", err)
+		rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusInternalServerError, err))
+		return
+	}
+
+	if LogLevel < 2 {
+		for _, req := range reqs {
+			log.Debug("HTTP Render %s - arch:%d archI:%d outI:%d aggN: %d", req, req.Archive, req.ArchInterval, req.OutInterval, req.AggNum)
+		}
+	}
+
+	out, err := s.getTargets(reqs)
+	if err != nil {
+		log.Error(3, "HTTP Render %s", err.Error())
+		rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusInternalServerError, err))
+		return
+	}
+
+	merged := mergeSeries(out)
+	sort.Sort(models.SeriesByTarget(merged))
+	defer func() {
+		for _, serie := range out {
+			pointSlicePool.Put(serie.Datapoints[:0])
+		}
+	}()
+
+	reqHandleDuration.Value(time.Now().Sub(pre))
+	rbody.WriteResponse(ctx, rbody.NewFastJsonResponse(200, models.SeriesByTarget(merged)))
+}
+
+func (s *Server) metricsFind(ctx *middleware.Context, request models.GraphiteFind) {
+	if ctx.OrgId == 0 {
+		rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusBadRequest, MissingOrgHeaderErr))
+		return
+	}
+
+	if request.Query == "" {
+		rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusBadRequest, MissingQueryErr))
+		return
+	}
+
+	if request.Format != "" && request.Format != "treejson" && request.Format != "json" && request.Format != "completer" {
+		rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusBadRequest, InvalidFormatErr))
+		return
+	}
+	// metricDefs only get updated periodically (when using CassandraIdx), so we add a 1day (86400seconds) buffer when
+	// filtering by our From timestamp.  This should be moved to a configuration option,
+	// but that will require significant refactoring to expose the updateInterval used
+	// in the MetricIdx.  So this will have to do for now.
+	if request.From != 0 {
+		request.From -= 86400
+	}
+	nodes, err := s.MetricIndex.Find(ctx.OrgId, request.Query, request.From)
+	if err != nil {
+		rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusBadRequest, err))
+		return
+	}
+
+	var b interface{}
+	switch request.Format {
+	case "", "treejson", "json":
+		b, err = findTreejson(request.Query, nodes)
+	case "completer":
+		b, err = findCompleter(nodes)
+	}
+
+	if err != nil {
+		rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusInternalServerError, err))
+		return
+	}
+	rbody.WriteResponse(ctx, rbody.NewJsonResponse(200, b, request.Jsonp))
+}
+
+func (s *Server) metricsIndex(ctx *middleware.Context) {
+	if ctx.OrgId == 0 {
+		rbody.WriteResponse(ctx, rbody.NewErrorResponse(http.StatusBadRequest, MissingOrgHeaderErr))
+		return
+	}
+
+	list := s.MetricIndex.List(ctx.OrgId)
+	rbody.WriteResponse(ctx, rbody.NewFastJsonResponse(200, models.MetricDefs(list)))
+}
+
+func parseTarget(target string) (string, string, error) {
+	var consolidateBy string
+	id := target
```

Comment:
also i see that the parseTarget is defined below where it is called.  I think it's a general good practice to keep the function definition above the place where it is called, when possible.  I mean i don't feel very strongly about it, but that's also how it is in other places in the code base